### PR TITLE
Fix bandwidth burst too large

### LINKED ISSUE #90838: Default bandwidth burst is too large
- Create a test pod with kubernetes.io/ingress-bandwidth and kubernetes.io/egress-bandwidth annotations
```
---
apiVersion: v1
kind: Pod
metadata:
  name: iperf3
  annotations:
    kubernetes.io/ingress-bandwidth: 10M
    kubernetes.io/egress-bandwidth: 10M
spec:
  containers:
  - name: iperf3
    image: networkstatic/iperf3
    args: ['-s']
```
- tc qdisc show
```
...
qdisc tbf 1: dev veth20261d7b root refcnt 2 rate 10Mbit burst 256Mb lat 25.0ms
qdisc ingress ffff: dev veth20261d7b parent ffff:fff1 ----------------
qdisc tbf 1: dev bwp201cb4fbd6ef root refcnt 2 rate 10Mbit burst 256Mb lat 25.0ms
```
- iperf3 test
```
[ ID] Interval           Transfer     Bandwidth       Retr  Cwnd
[  4]   0.00-1.00   sec   238 MBytes  1.99 Gbits/sec    0   8.12 MBytes
[  4]   1.00-2.00   sec  21.2 MBytes   178 Mbits/sec    0   8.12 MBytes
[  4]   2.00-3.00   sec  1.25 MBytes  10.5 Mbits/sec    0   8.12 MBytes
[  4]   3.00-4.00   sec  1.25 MBytes  10.5 Mbits/sec    0   8.12 MBytes
[  4]   4.00-5.00   sec  1.25 MBytes  10.5 Mbits/sec    0   8.12 MBytes
[  4]   5.00-6.00   sec  1.25 MBytes  10.5 Mbits/sec    0   8.12 MBytes
[  4]   6.00-7.00   sec  1.25 MBytes  10.5 Mbits/sec    0   8.12 MBytes
[  4]   7.00-8.00   sec  1.25 MBytes  10.5 Mbits/sec    0   8.12 MBytes
[  4]   8.00-9.00   sec   502 KBytes  4.11 Mbits/sec    0   8.12 MBytes
 
 
[  4]   9.00-10.00  sec  0.00 Bytes  0.00 bits/sec    0   8.12 MBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-10.00  sec   267 MBytes   224 Mbits/sec    0             sender
[  4]   0.00-10.00  sec   262 MBytes   220 Mbits/sec                  receiver
```
We can see that burst is too large, bandwidth limit is `10M` but burst can reach `1.99 Gbits/sec`.

### PR DESCRIPTION
**Pull request by liuxu623**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Fix bandwidth burst too large

**Which issue(s) this PR fixes**:

Fixes #90838

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/kubelet/dockershim/network/cni/cni.go [hunk 1]
```diff
@@ -425,7 +424,7 @@ func (plugin *cniNetworkPlugin) buildCNIRuntimeConf(podName string, podNs string
 		portMappingsCapability: portMappingsParam,
 	}
 
-	ingress, egress, err := bandwidth.ExtractPodBandwidthResources(annotations)
+	ingress, ingressBurst, egress, egressBurst, err := bandwidth.ExtractPodBandwidthResources(annotations)
```
Comment (mattjmcnaughton) [inline pkg/kubelet/dockershim/network/cni/cni.go:427]: Do we also need to update the usage of this method in `pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go`?

**Review by mattjmcnaughton — COMMENTED** [review]
/ok-to-test

I don't have a ton of context on the network specific components... And am also not positive about the pros/cons of using annotations for this information. I defer to a sig-network reviewer.

Can you also share more around what happens if we are not using `docker` as the container driver? Are these new annotations meaningless?

File: pkg/kubelet/dockershim/network/cni/cni.go (hunk 1 above)
Comment (liuxu623) [reply pkg/kubelet/dockershim/network/cni/cni.go:427]: yes

File: pkg/kubelet/dockershim/network/cni/cni.go [hunk 2]
```diff
@@ -436,11 +435,19 @@ func (plugin *cniNetworkPlugin) buildCNIRuntimeConf(podName string, podNs string
...
+				bandwidthParam.IngressBurst = int(ingress.Value())
+			}
 		}
 		if egress != nil {
 			bandwidthParam.EgressRate = int(egress.Value())
-			bandwidthParam.EgressBurst = math.MaxInt32 // no limit
+			if egressBurst != nil {
+				bandwidthParam.EgressBurst = int(egressBurst.Value())
+			} else {
+				bandwidthParam.EgressBurst = int(egress.Value())
```
Comment (johscheuer) [inline pkg/kubelet/dockershim/network/cni/cni.go:449]: This will break the default behaviour, since the bandwidth limitation is in a more or less experimental status this might be okay. We should at least ensure that the Release note is updated with a "Heads up" or an Action required. Also what is the pain point with letting `math.MaxInt32` the default when you can specify the burst rate now?

File: pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go [hunk 3]
```diff
@@ -410,7 +410,7 @@ func (plugin *kubenetNetworkPlugin) setup(namespace string, name string, id kube
 // TODO: replace with CNI traffic shaper plugin
 func (plugin *kubenetNetworkPlugin) addTrafficShaping(id kubecontainer.ContainerID, annotations map[string]string) error {
 	shaper := plugin.shaper()
-	ingress, egress, err := bandwidth.ExtractPodBandwidthResources(annotations)
+	ingress, _, egress, _, err := bandwidth.ExtractPodBandwidthResources(annotations)
```
Comment (johscheuer) [inline pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go:413]: Could you add here a `TODO` since these values should be used somewhere in the future (we can resolve this in an additional PR)

File: pkg/util/bandwidth/utils.go [hunk 4]
```diff
@@ -36,31 +36,53 @@ func validateBandwidthIsReasonable(rsrc *resource.Quantity) error {
...
+		}
+	}
+	str, found = podAnnotations["kubernetes.io/ingress-burst-bandwidth"]
+	if found {
+		ingressBurstValue, err := resource.ParseQuantity(str)
+		if err != nil {
+			return nil, nil, nil, nil, err
+		}
+		ingressBurst = &ingressBurstValue
+		if err := validateBandwidthIsReasonable(ingressBurst); err != nil {
```
Comment (johscheuer) [inline pkg/util/bandwidth/utils.go:61]: Do we need some additional validation here for the burst rate? e.g. from the [tc docs](http://man7.org/linux/man-pages/man8/tc-tbf.8.html) and we could validate that these values are reasonable.

```
If your buffer is too small, packets may be dropped because
more tokens arrive per timer tick than fit in your bucket.
The minimum buffer size can be calculated by dividing the rate
by HZ.
```

File: pkg/util/bandwidth/utils.go [hunk 5]
```diff
@@ -36,31 +36,53 @@ func validateBandwidthIsReasonable(rsrc *resource.Quantity) error {
...
-			return nil, nil, err
+			return nil, nil, nil, nil, err
 		}
 		egress = &egressValue
 		if err := validateBandwidthIsReasonable(egress); err != nil {
-			return nil, nil, err
+			return nil, nil, nil, nil, err
+		}
+	}
+	str, found = podAnnotations["kubernetes.io/egress-burst-bandwidth"]
```
Comment (johscheuer) [inline pkg/util/bandwidth/utils.go:76]: These blocks look like 4 times the same (expect for the annotation) do you think it make sense to refactor this into a single method and add some unit tests?

File: pkg/kubelet/dockershim/network/cni/cni.go [hunk 6]
```diff
@@ -436,11 +435,19 @@ func (plugin *cniNetworkPlugin) buildCNIRuntimeConf(podName string, podNs string
 			// https://github.com/containernetworking/plugins/blob/master/plugins/meta/bandwidth/README.md
 			// Rates are in bits per second, burst values are in bits.
 			bandwidthParam.IngressRate = int(ingress.Value())
-			bandwidthParam.IngressBurst = math.MaxInt32 // no limit
+			if ingressBurst != nil {
```
Comment (johscheuer) [inline pkg/kubelet/dockershim/network/cni/cni.go:438]: I think this should be part of `ExtractPodBandwidthResources` otherwise the behaviour for `kubenet` and `cni` plugins will differ?

Comment (liuxu623) [reply pkg/kubelet/dockershim/network/cni/cni.go:438]: kubenet  doesn't need set ingressBurst/egressBurst, kubenet and bandwidth plugin is different, I didn't use kubenet.

File: pkg/util/bandwidth/utils.go (hunk 4 above)
Comment (liuxu623) [reply pkg/util/bandwidth/utils.go:61]: We can make sure `ingressBurst` always greater `ingressRate`.

File: pkg/kubelet/dockershim/network/cni/cni.go (hunk 2 above)
Comment (liuxu623) [reply pkg/kubelet/dockershim/network/cni/cni.go:449]: I think we needn't specify burst rate in most situation, so let default burst equal rate is better.

File: pkg/kubelet/dockershim/network/cni/cni.go (hunk 6 above)
Comment (liuxu623) [reply pkg/kubelet/dockershim/network/cni/cni.go]: I change `ExtractPodBandwidthResources` to `ExtractPodBandwidthParam`, and make sure behaviour for kubenet and cni plugins are same.

**Comment by aojea** [comment]
just for reference,  previous attempt to add `burst` annotations was closed because @thockin asked for a KEP, I suggest to "convince" him first ;)

https://github.com/kubernetes/kubernetes/pull/87660#issuecomment-582008484

/assign @thockin

**Comment by liuxu623** [comment]
> just for reference [#87660 (comment)](https://github.com/kubernetes/kubernetes/pull/87660#issuecomment-582008484)
> , previous attempt to add `burst` annotations was closed because @thockin asked for a KEP, I suggest to "convince" him first ;)
> 
> /assign @thockin

or we can only change the default value，only modify two lines code
```
if ingress != nil {
  bandwidthParam.IngressRate = int(ingress.Value())
  bandwidthParam.IngressBurst = math.MaxInt32 // no limit
}
if egress != nil {
  bandwidthParam.EgressRate = int(egress.Value())
  bandwidthParam.EgressBurst = math.MaxInt32 // no limit
}
```
to
```
if ingress != nil {
  bandwidthParam.IngressRate = int(ingress.Value())
  bandwidthParam.IngressBurst = int(ingress.Value())
}
if egress != nil {
  bandwidthParam.EgressRate = int(egress.Value())
  bandwidthParam.EgressBurst = int(egress.Value())
}
```

**Comment by liuxu623** [comment]
@mattjmcnaughton @aojea @freehan Hi, I just change bandwidth burst value to equal rate, please help me review code, thank you.

**Comment by thockin** [comment]
This is effectively a breaking change.  Tell me why this is "better" than the previous, which wasn't set accidentally.

If we change this, we WILL impact someone who's depending on this behavior.

**Comment by liuxu623** [comment]
> This is effectively a breaking change. Tell me why this is "better" than the previous, which wasn't set accidentally.
> 
> If we change this, we WILL impact someone who's depending on this behavior.

The previous bandwidth limit doesn't work because burst is too large, you can see #90838, limit 10M but max bandwidth can reach `1.99 Gbits/sec`(node bandwidth limit), it's very dangerous in production.

**Comment by aojea** [comment]
> > This is effectively a breaking change. Tell me why this is "better" than the previous, which wasn't set accidentally.
> > If we change this, we WILL impact someone who's depending on this behavior.
> 
> The previous bandwidth limit doesn't work because burst is too large, you can see #90838, limit 10M but max bandwidth can reach `1.99 Gbits/sec`(node bandwidth limit), it's very dangerous in production.

To be honest, I tend to agree with @liuxu623 here, if I read it correctly https://lartc.org/howto/lartc.qdisc.classless.html#AEN691

>  burst/buffer/maxburst
> Size of the bucket, in bytes. This is the maximum amount of bytes that tokens can be available for instantaneously....
> 

a 256 Mb buffer just for interface seems excessive

> qdisc tbf 1: dev veth20261d7b root refcnt 2 rate 10Mbit burst 256Mb lat 25.0ms

**Comment by thockin** [comment]
I know precious little about the internals of tc.  Behaviorally this is a big change, but it's hard to say whether it is "safe" or not.  What if we set the burst to something like 2x the limit or something?  Do we even have a clear definition about what burst means?

@dcbw since CNI?  I don't feel equipped to answer this without a bunch of experimenting.

**Comment by jayunit100** [comment]
Sounds like we need to discuss this @ sig-net next wk @liuxu623 if you can come?

**Comment by jayunit100** [comment]
I wonder if https://github.com/kubernetes/enhancements/pull/866 (moving Dockershim out of tree) implies that any changes to Dockershim which aren't mission critical should be avoided, for now.

File: pkg/kubelet/dockershim/network/cni/cni_test.go [hunk 7]
```diff
@@ -304,7 +304,7 @@ func TestCNIPlugin(t *testing.T) {
 	}
 	expectedBandwidth := map[string]interface{}{
 		"ingressRate": 1000000.0, "egressRate": 1000000.0,
-		"ingressBurst": 2147483647.0, "egressBurst": 2147483647.0,
+		"ingressBurst": 1000000.0, "egressBurst": 1000000.0,
```
Comment (jayunit100) [inline pkg/kubelet/dockershim/network/cni/cni_test.go:307]: Good catch ! but , i  think changing this is the sticking point, which the sig moving the docker provider out of tree will have to own

**Comment by jayunit100** [comment]
We discussed this in sig-net today - 

- there are no strong opinions for or against this.
- the semantics around what this burst should do and how it's implemented are pretty complicated.
- its not out of the question to merge this specific fix for the docker shim side of things.

@liuxu623 can you provide us some intuition on how this effects you in production.  

In general, does anyone else have any feedback on fixing the bandwidth burstyness in production?

**Review by thockin — COMMENTED** [review]
I have been chewing on this since sig-net yesterday.  It seems like such a small thing, but the risk of explosion is real. 

The behavior of this was never really specified, but this interpretation seems reasonable.  That said, it seems the CNI plugin (https://github.com/containernetworking/plugins/blob/master/plugins/meta/bandwidth/main.go) doesn't make this assumption either, so fixing it here would still be "broken" in CNI.

Given that, my inclination is just to close this as infeasible.

The alternative would be to deprecate the old, add a new frontend API to configure these parameters more explicitly.

@squeed @dcbw ?

**Comment by squeed** [comment]
It looks like only kubenet gets this right, by dumb luck

TL;DR - we should merge this, but we need to fix it in CNI.

The CNI convention tries to mirror kubenet, but it looks like we compute burst completely wrong in dockershim/CNI. I suspect this behavior is replicated in ocicni and go-cni, so that means it needs to be fixed thrice. (Aside: once I'm off parental leave, I'll put some feelers out for writing an official sig-net cni driver)

Kubenet relies on the default of `tc class add htb rate XXX`, which seems reasonable. The default CNI plugin uses `tbf`, which needs a more subtle burst calculation, and has no defaulting in the `tc` binary.

Given that the meaning of burst depends on the **specific qdisc in use**, we should ultimately

* Stop requiring burst in the CNI convention / annotations
* Require that plugins compute a reasonable value for burst if not supplied

**Comment by squeed** [comment]
Now that I've had time to think about this, here's what we should do:

1. Deprecate the existing **CNI** burst parameter, given that it's poorly defined but currently in use.
2. Define a new parameter, `burstPercent`, which is a multiple of the bandwidth rate that the container is allowed to use instantaneously. The plugin would be responsible for computing this value.

We shouldn't have to change the existing pod annotations.

If neither burst parameter is provided, then a `burstPercent` of 100 should be assumed.

Thoughts, @dcbw @thockin?

**Comment by jayunit100** [comment]
fyi from an email sent out this morning, looks like from sig-nodes's perspective, dockershim may be removed from the next few releases anyways.  is it really worth modifying at this point?

**Comment by squeed** [comment]
@jayunit100 I think so, yes - the annotations as defined by dockershim are useable; whereas the same annotations as interpreted by cni are unreasonable and poorly specified.

**Comment by jayunit100** [comment]
Seems like this is in a state of "logically valid bug, but not something anyone cares about that much" ...
